signac/core/synceddict.py

```diff
@@ -128,10 +128,13 @@ def _dfs_convert(self, root):
         elif type(root) in (list, tuple):
             return _SyncedList(root, parent=self)
         elif NUMPY:
-            if isinstance(root, numpy.number):
+            if isinstance(root, numpy.ndarray):
+                if root.shape == ():
+                    return root.item()
```

Comment:
For a unyt.quantity, this strips the unit value.
```
import unyt as u
project = signac.init_project("debug")  
job = project.open_job({"mass": 1.0 * u.gram})
job.sp["mass"]
```
is 1.0 of type float
Would there be an issue with 
```
                if root.shape == ():
                    return root
```
?

Reply:
Yes, this is an unfortunate limitation of the JSON backend and is basically the same reason why the lines below convert `numpy.ndarray` into a JSONList:
```
>>> import numpy as np
>>> job = project.open_job({'a': np.array([1])})
>>> job
Job(project=Project.get_project('/Users/vyasramasubramani/tmp2'), statepoint={'a': [1]})
>>> job.sp.a
[1]
>>> type(job.sp.a)
<class 'signac.core.synced_collections.collection_json.JSONList'>
```
There's no way to store a numpy array in JSON (just try doing `json.dumps(numpy.array([1]))` and you'll see that it fails), so we have to convert it into something like a list. I previously raised errors like this as a possible reason not to automatically convert and force users to do it up front (see [this thread](https://github.com/glotzerlab/signac/issues/380#issuecomment-681178761)), but the upshot is that there's no easy way to support a faithful representation of objects like numpy arrays in JSON. unyt ends up falling in the same bucket.

Reply:
@jennyfothergill: @vyasr's comment is accurate, signac requires state points to be JSON-encodable, which limits us to basic types like `dict`, `list`, `float`, `int`, `str`. If you need to store unyt quantities to disk, you may want to adopt one of the methods suggested in the unyt docs: https://unyt.readthedocs.io/en/stable/usage.html#writing-data-with-units-to-disk If you need to use a unyt quantity in a job state point or document, you might be better off writing your own wrapper that converts quantities to something like `{"value": 1.0, "unit": "gram"}` (that's a JSON-encodable dictionary).

Reply:
OK this solution `{"value": 1.0, "unit": "gram"}` will work. Should there be some sort of warning for unyt.quantities? I don't know how commonly they're used.

Reply:
Since numpy arrays are the only object that we currently convert, what I would suggest is that we throw a warning (using `warnings.warn`) whenever the `_SyncedDict` or `JSONDict` detect something that is a numpy array. @bdice thoughts? I've advocated that before as well, I think it's helpful to the user to indicate when we're converting something that might be unexpected.

Reply:
This discussion is closely related to #380. I don't really know what to do, we've been back and forth on "transparent" NumPy support for a long time. It's fine with me if someone wants to make that PR to add a warning.

Reply:
@bdice have we discussed this anywhere else? I don't recall any discussion about the viability of transparent numpy support outside of #380, which is where I suggested adding the warning. Is there any other context that I'm missing with respect to this decision? If not, I'd suggest that @jennyfothergill just add the warning here and we can move on.